test/system/kubernetes/fluentBitRotater.sh

```diff
@@ -0,0 +1,261 @@
+# (where <logname> can be expanded into <namespace>.<pod_name-tag>.<container_name>.<container_id>)
+# in this structure:
+#
+#   namespace/
+#   └── pod_name/
+#       └── <tag-container_name>-<substr(container_id, 0, 7)>.log-<epoch>.gz
+#
+# Todo: Return list of files moved to use for 'epoch_to_utc'.
+redirect() {
+    for src in $@; do
+        local ifs=$IFS
+        IFS='_'
+        local log=${src%.log*}
+        local ext=${src#$log.}
+        log=${log#kube.var.log.containers.}
+        read pod namespace container <<< "$log"
+        local container_id=${container##*-}
+        local container_name=${container%-*}
+        local tag=${pod##*-}
+        pod=${pod%-$tag}
+        local dest="$namespace/$pod/$tag-$container_name-$(printf "%.8s" "$container_id").$ext"
+        IFS=$ifs
+        mkdir -p "$(dirname "$dest")"
+        mv "$src" "$dest"
+    done
+}
+
+rotate() {
+    debug "Attempting to acquire /var/lock/rotate.lock ..."
+    (
+        if ! flock 200; then
+            return 1
+        fi
+        if [  "$PWD" != "$MOUNT_PATH" ]; then
+            cd "$MOUNT_PATH"
+        fi
+        opts=${1:-""}
+        start_time=$(date +%s%3N)
+        if [ "$opts" = "--force" ]; then
+            debug "Received --force logrotate option."
+        fi
+        if ! ls *.log > /dev/null 2>&1; then
+            debug "No log files found -- skipping rotation."
+            return 0
+        fi
+        debug "Running logrotate."
+        if ! logrotate "$LOG_ROTATE_CONF_PATH" $opts; then
+            return 2
+        fi
+        local files=$(find . -maxdepth 1 -regextype posix-extended -regex '.*\.log$' -printf "%T@ %P\n" | sort -n | grep -oE '[^ ]*$')
+        debug "Found $(items $files) '.log' files."
+        if [ -n "$files" ]; then
+            log "Looking for stale log files."
+        fi
+        orphans "$files"
+        # 'orphans' can create newly compressed files, so we must update state.
+        files=$(find . -maxdepth 1 -regextype posix-extended -regex ".*\.$LOG_EXT$" -printf "%T@ %P\n")
+        files=$(echo "$files" | sort -n | grep -oE '[^ ]*$')
+        debug "Redirecting $(items $files) recently compressed '.$LOG_EXT' files."
+        redirect "$files"
+        ## Rotated but not renamed (i.e. the files that were *just* rotated).
+        rotated=$(find . -mindepth 2 -regextype posix-extended -regex "\.\/.*[0-9]{10}\.$LOG_EXT$" -printf "%p\n")
+        debug "Found $(items $rotated) recently rotated '.log' files."
+        for name in $rotated; do
+            epoch_to_utc "$name"
+        done;
+        # Attempt to reclaim any old log files that have been 'redirected' (those not just rotated).
+        reclaim
+        end_time=$(date +%s%3N)
+        if [ -n "$rotated" ]; then
+            log "Rotation cycle completed in $((end_time-start_time)) milliseconds."
+        fi
+        echo ""
+    ) 200> /var/lock/rotate.lock 
+}
+
+watch() {
```

Comment:
Who is taking care of starting and monitoring this process? If it crashes, I'm afraid that the PVC with logs will get filled, right?

Reply:
This process is started by a [deployment](https://github.com/pravega/pravega/blob/041d2ba9582aaa65dd9fb8046554163053290000/test/system/kubernetes/fluentBitSetup.sh#L472) so if it crashes it will at least be restarted. If there's a fundamental issue with the logic and it keeps restarting before finishing any reclamation the PVC would overflow yeah. 

I haven't tested it really in a 'hostile' environment so I would want to follow this PR up and make sure that a crash does not prevent successive restarts from failing.